airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLBulkLoadHandler.kt

```diff
@@ -0,0 +1,244 @@
+     * @param rowsPerBatch Optional parameter to control the batching size in BULK INSERT.
+     */
+    fun bulkLoadAndUpsertForDedup(
+        primaryKeyColumns: List<String>,
+        nonPkColumns: List<String>,
+        dataFilePath: String,
+        formatFilePath: String,
+        rowsPerBatch: Long? = null
+    ) {
+        if (primaryKeyColumns.isEmpty()) {
+            throw IllegalArgumentException("At least one primary key column is required.")
+        }
+
+        val tempTableName = generateLocalTempTableName()
+
+        dataSource.connection.use { conn ->
+            conn.autoCommit = false
+            try {
+                // Create the temp table
+                createTempTable(conn, tempTableName)
+                // Bulk load the CSV data into the temp table
+                val bulkInsertSql =
+                    buildBulkInsertSql(
+                        fullyQualifiedTableName = tempTableName,
+                        dataFilePath = dataFilePath,
+                        formatFilePath = formatFilePath,
+                        rowsPerBatch = rowsPerBatch,
+                    )
+                logger.info {
+                    "Starting bulk insert into temp table: $tempTableName from file: $dataFilePath"
+                }
+                conn.prepareStatement(bulkInsertSql).use { stmt -> stmt.executeUpdate() }
+                logger.info { "Bulk insert completed successfully for temp table: $tempTableName" }
+
+                // Merge into the main table
+                val mergeSql = buildMergeSql(tempTableName, primaryKeyColumns, nonPkColumns)
+                logger.info { "Starting MERGE into: $schemaName.$mainTableName" }
+                conn.prepareStatement(mergeSql).use { stmt -> stmt.executeUpdate() }
+                logger.info {
+                    "MERGE completed successfully into table: $schemaName.$mainTableName"
+                }
+
+                handleCdcDeletes(conn)
+                conn.commit()
+            } catch (ex: SQLException) {
+                logger.error(ex) {
+                    "Error during bulk load & upsert; rolling back. Cause: $ex.message"
+                }
+                conn.rollback()
+                throw ex
+            }
+        }
+    }
+
+    private fun handleCdcDeletes(conn: Connection) {
+        if (mssqlQueryBuilder.hasCdc) {
+            logger.info {
+                "Starting removal of deleted records in table: $schemaName.$mainTableName"
+            }
+            mssqlQueryBuilder.deleteCdc(conn)
+            logger.info {
+                "Deleted records removal completed successfully in table: $schemaName.$mainTableName"
+            }
+        }
+    }
+
+    private fun quoteIdentifier(identifier: String): String {
+        // Split on '.', then bracket each piece (like database.schema.table ->
+        // [database].[schema].[table])
+        return identifier.split(".").joinToString(".") { segment ->
```

Comment:
feels dangerous to split on `"."` for quoting, I suspect this could break for identifiers with a `.`.